### PR TITLE
add a directory prefix `gitea-src-VERSION` to release-tar-file

### LINKED ISSUE #19066: Tarbomb in release src tarball file
### Gitea Version

1.16.3

### Git Version

N/A

### Operating System

N/A

### How are you running Gitea?

tar xf gitea-src-1.16.3.tar.gz

### Database

_No response_

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

N/A

### Description

Gitea official src, gitea-src-1.16.3.tar.gz, (from releases page at [github](https://github.com/go-gitea/gitea/releases), particulary at [release of 1.16.3](https://github.com/go-gitea/gitea/releases/tag/v1.16.3)) contains [tarbomb](https://en.wikipedia.org/wiki/Tar_(computing)#Tarbomb). Such behavior is considered bad etiquette on the part of the archive's creator.

Expected behavior is (after untarring) to find this pile of files in subdir named gitea-src-1.16.3 or even better in subdir named gitea-1.16.3.

### Screenshots

N/A

### PR DESCRIPTION
**Pull request by jklippel**
Do not package tar file such that the sources are unpacked into the
current working directory. Use a directory prefix instead on creating
the tar file.

Fixes: #19066

## :warning: BREAKING :warning: 

Prior to this PR the gitea src tar.gz would extract to the current directory. Users requiring the previous behaviour should use tar options as appropriate.

### CODE REVIEW

**Comment by techknowlogick** [comment]
Thanks for this PR, could you upload a tar generated with the above? I'd love to pass it to downstream folk (nixos, FreeBSD/OpenBSD,and others) who package gitea using this generated file so we can ensure that if it breaks their packaging it can be resolved prior to a release.

**Comment by wxiaoguang** [comment]
Dup with:
* https://github.com/go-gitea/gitea/pull/19153

**Comment by jklippel** [comment]
Packaged tag v1.16.5 by applying the proposed changes to the Makefile.
[gitea-src-1.16.5.tar.gz](https://github.com/go-gitea/gitea/files/8488383/gitea-src-1.16.5.tar.gz)

**Comment by techknowlogick** [comment]
@jklippel thanks for uploading that sample, I've reached out to the *BSD folk and am awaiting feedback, as for nix I've done some preliminary testing and the build will need some changes if/when this is merged.
 
@wxiaoguang I think this is the better PR because it looks for bsdtar vs tar.

cc: @Ma27 @kolaente (as other nix packagers for gitea)

**Comment by wxiaoguang** [comment]
> @wxiaoguang I think this is the better PR because it looks for bsdtar vs tar.

However it does nothing with bsdtar, there is `-s` for bsdtar to add the prefix.

Either we stop with an error when there is bsdtar, or we use bsdtar with `-s`. Otherwise there will be inconsistent output on different systems.

**Comment by jklippel** [comment]
> > @wxiaoguang I think this is the better PR because it looks for bsdtar vs tar.
> 
> However it does nothing with bsdtar, there is `-s` for bsdtar to add the prefix.
> 
> Either we should stop the make with an error when there is bsdtar, or we should use bsdtar with `-s` to make the output has the same prefix. Otherwise there will be inconsistent output on different systems.

I'm in favour of using `-s`. I will look into it. Is there any specific version of bsdtar or *bsd this should work on?

**Comment by wxiaoguang** [comment]
I am not a BSD user 😂  I think we can assume that almost all bsdtar have the `-s` option.

**Comment by jklippel** [comment]
[gitea-src-1.16.5.tar.gz](https://github.com/go-gitea/gitea/files/8491747/gitea-src-1.16.5.tar.gz)
Implemented -s for bsd tar
Implemented and tested using freebsd 12.

**Comment by kevans91** [comment]
> Thanks for this PR, could you upload a tar generated with the above? I'd love to pass it to downstream folk (nixos, FreeBSD/OpenBSD,and others) who package gitea using this generated file so we can ensure that if it breaks their packaging it can be resolved prior to a release.

FWIW- I don't maintain FreeBSD's port, but this change will in-fact simplify the port slightly by aligning to what the ports framework typically expects from a distfile. This is likely a net -1 line change for us, and not one that's going to cause any problems.

**Review by wxiaoguang — APPROVED** [review]

**Comment by silverwind** [comment]
> Is there any specific version of bsdtar or *bsd this should work on?
> I am not a BSD user 😂 I think we can assume that almost all bsdtar have the -s option.

The version on macOS should be supported for release creation and it does not have `-s`. FreeBSD tar indicates it does [not have it](https://www.freebsd.org/cgi/man.cgi?query=bsdtar&sektion=1&manpath=FreeBSD+6.0-RELEASE) either.

If everything fails, we could just require GNU tar on macOS via`gtar`, available via `brew install gnu-tar`. Thought I would prefer a solution that would work with regular bsdtar, why not just `mkdir`and move files before taring them?

**Comment by wxiaoguang** [comment]
> The version on macOS should be supported for release creation and it does not have `-s`. FreeBSD tar indicates it does [not have it](https://www.freebsd.org/cgi/man.cgi?query=bsdtar&sektion=1&manpath=FreeBSD+6.0-RELEASE) either.
> 

macOS tar does have the `-s`, check your `man tar` please.


> If everything fails, we could just require GNU tar on macOS via`gtar`, available via `brew install gnu-tar`. Thought I would prefer a solution that would work with regular bsdtar, why not just `mkdir`and move files before taring them?

Why are you reading a FreeBSD 6.0 document ....... FreeBSD 12 does have it. https://www.freebsd.org/cgi/man.cgi?query=bsdtar&apropos=0&sektion=1&manpath=FreeBSD+12.0-RELEASE&arch=default&format=html

**Comment by silverwind** [comment]
Ah, I missed it `man bsdtar` actually has it on macOS.

```
bsdtar 3.5.1 - libarchive 3.5.1 zlib/1.2.11 liblzma/5.0.5 bz2lib/1.0.8
```
```
     -s pattern
             Modify file or archive member names according to pattern.  The pattern has
             the format /old/new/[ghHprRsS] where old is a basic regular expression, new
             is the replacement string of the matched part, and the optional trailing
             letters modify how the replacement is handled.  If old is not matched, the
             pattern is skipped.  Within new, ~ is substituted with the match, \1 to \9
             with the content of the corresponding captured group.  The optional trailing
             g specifies that matching should continue after the matched part and stop on
             the first unmatched pattern.  The optional trailing s specifies that the
             pattern applies to the value of symbolic links.  The optional trailing p
             specifies that after a successful substitution the original path name and
             the new path name should be printed to standard error.  Optional trailing H,
             R, or S characters suppress substitutions for hardlink targets, regular
             filenames, or symlink targets, respectively.  Optional trailing h, r, or s
             characters enable substitutions for hardlink targets, regular filenames, or
             symlink targets, respectively.  The default is hrs which applies
             substitutions to all names.  In particular, it is never necessary to specify
             h, r, or s.
```

> Why are you reading a FreeBSD 6.0 document

It was the first result that came up in google. It's fine if the argument was added in later versions.

**Review by zeripath — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -646,7 +646,10 @@ release-sources: | $(DIST_DIRS)
 	echo $(VERSION) > $(STORED_VERSION_FILE)
 # bsdtar needs a ^ to prevent matching subdirectories
 	$(eval EXCL := --exclude=$(shell tar --help | grep -q bsdtar && echo "^")./)
-	tar $(addprefix $(EXCL),$(TAR_EXCLUDES)) -czf $(DIST)/release/gitea-src-$(VERSION).tar.gz .
+# use transform to a add a release-folder prefix; in bsdtar the transform parameter equivalent is -s
+	$(eval TRANSFORM := $(shell tar --help| grep -q bsdtar || echo "--transform 's|^./|gitea-src-$(VERSION)/|'"))
+	$(eval TRANSFORM := $(shell tar --help| grep -q bsdtar && echo "-s '/^./gitea-src-$(VERSION)/'"))
```
Comment (silverwind) [inline Makefile]:
```suggestion
	$(eval TRANSFORM := $(shell tar --help | grep -q bsdtar && echo "-s '/^./gitea-src-$(VERSION)/'" || echo "--transform 's|^./|gitea-src-$(VERSION)/|'"))
```

I think we can do this in one `tar` call.

**Comment by jklippel** [comment]
Attached files built with branch and latest commit (93e4cf8):
[freebsd-gitea-src-1.16.5.tar.gz](https://github.com/go-gitea/gitea/files/8546709/freebsd-gitea-src-1.16.5.tar.gz)
[gitea-src-1.16.5.tar.gz](https://github.com/go-gitea/gitea/files/8546710/gitea-src-1.16.5.tar.gz)

**Pull request merged**